### PR TITLE
Set precedence to DATABASE_URL over POSTHOG_DB_* fields

### PR DESCRIPTION
**Pull request by neilkakkar**
## Changes

This reconciles precedence order to be the same as on PostHog/posthog. This is important, because otherwise cases are possible where the Django app connects to postgres without issues, but the plugin server doesn't: so events are shown on the UI, but aren't ingested properly.

Specially, have seen this happen with SSL enabled DB urls, like in this case: https://posthogusers.slack.com/archives/C01GLBKHKQT/p1621416865003200

## Checklist

-   [ ] Updated Settings section in README.md, if settings are affected
-   [ ] Jest tests

### CODE REVIEW

File: src/utils/db/hub.ts [hunk 1]
```diff
@@ -135,7 +135,8 @@ export async function createHub(
         timeStr ? DateTime.fromSQL(timeStr, { zone: 'utc' }).toISO() : null
     )
 
-    const postgres = createPostgresPool(serverConfig)
+    const configOrDatabaseUrl = serverConfig.DATABASE_URL ? serverConfig.DATABASE_URL : serverConfig
```
Comment (Twixes) [inline src/utils/db/hub.ts:138]: This will always result in `DATABASE_URL` being used, because it has a non-empty default value (while `POSTHOG_DB_NAME` is null by default).

**Review by Twixes — CHANGES_REQUESTED** [review]
This should not be needed for any setup though 🤔 `POSTHOG_DB_NAME` – AFAIK – has to be explicitly set by the user, in which case we can assume it's more deliberate than `DATABASE_URL` – which practically always has a value, because it has a default.

File: src/utils/db/hub.ts (hunk 1 above)
Comment (neilkakkar) [reply src/utils/db/hub.ts:138]: Oh, right, this is annoying :$ - looks like it's not required on main PostHog: https://github.com/PostHog/posthog/blob/master/posthog/settings.py#L369-L376 - and hence the precedence order there.

If we use `POSTHOG_DB_NAME`, while PostHog/posthog uses DATABASE_URL, that's.. not good.

I wonder if we can get around this by removing the default value for `DATABASE_URL` ? Does that break things / workflows elsewhere? (I'm guessing probably yes)

.. or maybe it's a documentation thing? We explicitly mention to use one or the other? Although, I'd prefer to fix this ourselves so things are consistent.

Comment (Twixes) [reply src/utils/db/hub.ts:138]: Yeah, def not good to have a discrepancy. I suppose the fix would then be removing the default value of `DATABASE_URL` here in the non-test non-dev case, and adjusting `createPostgresPool` for same control flow as in the main repo.

File: src/utils/utils.ts [hunk 2]
```diff
@@ -416,22 +416,27 @@ export function createPostgresPool(
...
                   connectionString: configOrDatabaseUrl,
               }
-            : configOrDatabaseUrl.POSTHOG_DB_NAME
+            : configOrDatabaseUrl.DATABASE_URL
             ? {
-                  database: configOrDatabaseUrl.POSTHOG_DB_NAME,
+                  connectionString: configOrDatabaseUrl.DATABASE_URL,
+              }
+            : {
+                  database: configOrDatabaseUrl.POSTHOG_DB_NAME || undefined,
```
Comment (neilkakkar) [inline src/utils/utils.ts:434]: Is there a better way to convert `null` to `undefined` for types?

Comment (Twixes) [reply src/utils/utils.ts:434]: You can also use `??` which is the _nullish_ coalescing operator – where `||` returns right side when left side is _falsy_, `??` returns right side only when left side is `null` or `undefined`. So depends on whether we want the plugin server to treat `''` as a valid (if empty) string value (→ `??`), or as a lack of value i.e. `undefined` (→ `||`).

**Review by Twixes — COMMENTED** [review]
Nice, discrepancy is no more. A few extra edge cases

File: tests/helpers/sql.ts [hunk 3]
```diff
@@ -33,7 +33,7 @@ export async function resetTestDatabase(
     extraRows: ExtraDatabaseRows = {}
 ): Promise<void> {
     const config = { ...defaultConfig, ...extraServerConfig }
-    const db = new Pool({ connectionString: config.DATABASE_URL })
+    const db = new Pool({ connectionString: config.DATABASE_URL || undefined })
```
[the remark this answers is not in the document]
Comment (neilkakkar) [reply tests/helpers/sql.ts:36]: Fair

File: tests/helpers/graphile.ts [hunk 4]
```diff
@@ -4,7 +4,7 @@ import { Pool } from 'pg'
 import { PluginsServerConfig } from '../../src/types'
 
 export async function resetGraphileSchema(serverConfig: PluginsServerConfig): Promise<void> {
-    const graphileUrl = serverConfig.JOB_QUEUE_GRAPHILE_URL || serverConfig.DATABASE_URL
+    const graphileUrl = serverConfig.JOB_QUEUE_GRAPHILE_URL || serverConfig.DATABASE_URL || undefined
```
[the remark this answers is not in the document]
Comment (neilkakkar) [reply tests/helpers/graphile.ts:7]: Can't be `serverConfig`, since it's directly going into the pool, how about I use the default it was supposed to be for tests? (since this is in tests?)

[the remark this answers is not in the document]
Comment (neilkakkar) [reply tests/helpers/graphile.ts:7]: Err, I meant using the non-null assertion operator

File: src/utils/utils.ts (hunk 2 above)
Comment (neilkakkar) [reply src/utils/utils.ts:434]: Will do what PostHog main does: null is not ok, falsy is ok.

File: tests/helpers/graphile.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Twixes) [reply tests/helpers/graphile.ts]: Ah, right, I missed that this is a helper for tests. In that case yep, good to just assert non-null

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -118,6 +118,8 @@ jobs:
                   echo "⏳ Waiting 10 seconds to see if it runs"
                   sleep 10
 
+                  cat tmp/plugin-server.log
```
Comment (Twixes) [inline .github/workflows/ci.yml:121]: This way we have this line once always and also a second time if the plugin server failed somehow (down below) 🤔

**Review by Twixes — APPROVED** [review]

File: .github/workflows/ci.yml (hunk 5 above)
Comment (neilkakkar) [reply .github/workflows/ci.yml]: <Sigh>, Thanks!
